File: cli/layout/image_to_text.py
from PIL import Image

ESCAPE = "\033["


def rgb(red: int, green: int, blue: int):
    return ESCAPE + "38;2;" + str(red) + ";" + str(green) + ";" + str(blue) + "m"


def image_to_text(path: str, scale: float, super_scale=False):
    img = Image.open(path)
    image = img.resize(
        (int(img.height * scale), int(scale * img.width // 3)), Image.NEAREST
    )
    out = ""
    for x in range(0, image.height):
        for y in range(0, image.width):
            colors = image.getpixel((y, x))
            out += rgb(colors[0], colors[1], colors[2])
            if len(colors) > 3:
                if colors[3] > 0.7:
                    out += "█"
                elif colors[3] > 0.2:
                    out += "#"
                else:
                    out += "."
            else:
                out += "█"
        out += "\n"
    return out

File: src/layout/image_to_text.rs
extern crate image;

use std::path::Path;

use terminal::color::rgb;
use image::{imageops, GenericImageView};

pub fn ascii_image(input_path: &str, scale: f64) -> crate::Result<String> {
    let img = image::open(Path::new(input_path)).map_err(|_e| "Failed to open image")?;
    let img_width = img.dimensions().0;
    let img_height = img.dimensions().1;
    #[allow(clippy::cast_sign_loss)]
    let new_img = img.resize(
        (f64::from(img_width) * scale) as u32,
        (f64::from(img_height) * scale) as u32,
        imageops::FilterType::Nearest,
    );
    let mut text = String::new();
    for p in new_img.pixels() {
        if p.1 == 0 {
            text += "\n";
        }
        let pixel = p.2 .0;
        text += &rgb(pixel[0], pixel[1], pixel[2]);
        text += "█";
    }
    Ok(text)
}
